include/basefunctions.hxx

```diff
@@ -141,6 +141,162 @@ inline ROOT::RDF::RNode DefineQuantity(ROOT::RDF::RNode df,
                                        T const &value) {
     return df.Define(outputname, [value]() { return value; }, {});
 }
+
+/// Build a new column with a collection of Lorentz vectors per dataframe entry, which are created
+/// from the pt, eta, phi and mass columns of a collection.
+///
+/// For instance, this can be used to create four-vector objects from the four-vector component
+/// columns of a NanoAOD collection.
+///
+/// The function expects pt, eta phi and mass columns, which contain `ROOT::VecOps::RVec<float>`
+/// objects. In addition, the argument `collection_index` must point to a column which contains
+/// index lists of type `ROOT::VecOps::RVec<int>`. These index lists contain the indices of
+/// elements, for which four-vectors are built. The output column contains collections of
+/// four-vectors `ROOT::VecOps::RVec<PtEtaPhiMVector>`. The output collection only contains
+/// the elements, which have been selected in the `collection_index` list.
+///
+/// \param df Input dataframe
+/// \param outputname name of the output column
+/// \param pts column name with the pt values of the collection's objects
+/// \param etas column name with the eta values of the collection's objects
+/// \param phis column name with the phi values of the collection's objects
+/// \param masses column name with the mass values of the collection's objects
+/// \param collection_index column name for index lists of the elements for which four-vectors are built
+///
+/// \returns a dataframe with the new column
+inline ROOT::RDF::RNode BuildP4Collection(ROOT::RDF::RNode df,
+                                          const std::string &outputname,
+                                          const std::string &pts,
+                                          const std::string &etas,
+                                          const std::string &phis,
+                                          const std::string &masses,
+                                          const std::string &collection_index) {
+    auto build_collection_p4 = [](const ROOT::RVec<float> &pts,
+                                  const ROOT::RVec<float> &etas,
+                                  const ROOT::RVec<float> &phis,
+                                  const ROOT::RVec<float> &masses,
+                                  const ROOT::RVec<int> &collection_index) {
+        auto p4s = ROOT::RVec<ROOT::Math::PtEtaPhiMVector>();
+        for (int i = 0; i < collection_index.size(); ++i) {
+            auto p4 = ROOT::Math::PtEtaPhiMVector(pts.at(i), etas.at(i), phis.at(i), masses.at(i));
+            p4s.push_back(p4);
+        }
+        return p4s;
+    };
+    return df.Define(outputname, build_collection_p4, {pts, etas, phis, masses, collection_index});
+}
+
```

Comment:
There is https://root.cern/doc/master/group__vecops.html#ga84ffef84f7829e9d904cd92a57a0cd9a to further shorten this code and eliminate the for-loop